### PR TITLE
[LW-10042] chore: accept older minor/patch versions in routes

### PR DESCRIPTION
**Pull request by michalrus**
[LW-10042](https://input-output.atlassian.net/browse/LW-10042)

# Context

An alternative to #1140, IMO more correct.

# Proposed Solution

Change the `version.json` file instead.

# Important Changes Introduced

n/a

[LW-10042]: https://input-output.atlassian.net/browse/LW-10042?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: nix/cardano-services/deployments/backend-ingress.nix
Comment (michalrus) [inline nix/cardano-services/deployments/backend-ingress.nix:71]: @gytis-ivaskevicius should `max(version)` be used here? Or all of them?

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (mkazlauskas) [inline packages/cardano-services-client/scripts/createVersionSource.js:36]: If a major version is bumped, then this server doesn't support previous major versions. I suggest to also rename the json file to `supportedVersions.json`
```suggestion
      versionsData[key] = [...versionsData[key].filter(sameMajorVersion(apiVersion[key])), apiVersion[key]];
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
Looks good, just one change request below.
Also I think this needs some manual testing (bump some minor version in openApi.json and re-generate, bump some major version and re-generate etc.).

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (gytis-ivaskevicius) [inline packages/cardano-services-client/scripts/createVersionSource.js:38]:
```suggestion
  Object.entries(apiVersion).forEach(({key, value}) => {
    const versionsValue = versionsData[key] ?? [ ];
    versionsData[key] = !versionsValue.includes(apiVersion[key]) 
      ? [...versionsValue, value]
      : versionsValue
  })
```

**Review by gytis-ivaskevicius — APPROVED** [review]

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (michalrus) [reply packages/cardano-services-client/scripts/createVersionSource.js:38]: I wrote that first, but linter rules in this repository forbid `Object…forEach` :eyes:

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (gytis-ivaskevicius) [reply packages/cardano-services-client/scripts/createVersionSource.js:36]: Can we somehow fix up openapi definitions to just include versions that service supports?

File: nix/cardano-services/deployments/backend-ingress.nix
Comment (michalrus) [reply nix/cardano-services/deployments/backend-ingress.nix:71]: Ah, no, `max(version)` was pseudocode, I called this construct (`lib.last (lib.sort lib.versionOlder`) `max` :)

Comment (gytis-ivaskevicius) [reply nix/cardano-services/deployments/backend-ingress.nix:71]: Ah, im okay with either

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (gytis-ivaskevicius) [reply packages/cardano-services-client/scripts/createVersionSource.js:38]: :facepalm:

Comment (gytis-ivaskevicius) [reply packages/cardano-services-client/scripts/createVersionSource.js:38]: I think something like this should work too
```suggestion
  for (const {key, value} of Object.entries(apiVersion)) {
    const versionsValue = versionsData[key] ?? [ ];
    versionsData[key] = !versionsValue.includes(apiVersion[key]) 
      ? [...versionsValue, value]
      : versionsValue
  }
```

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (mkazlauskas) [reply packages/cardano-services-client/scripts/createVersionSource.js:36]: It's already doing that by using semantic version :)

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (mkazlauskas) [reply packages/cardano-services-client/scripts/createVersionSource.js:25]: Same as https://github.com/input-output-hk/cardano-js-sdk/pull/1146#discussion_r1522952427 - don't see a reason to add any extra data to openApi.json. `major.minor.patch` provides enough information to determine supported versions

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (michalrus) [reply packages/cardano-services-client/scripts/createVersionSource.js]: Fixed in 99e3c4d91fc6f7079a2509553b3e3977efd234d5

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (michalrus) [reply packages/cardano-services-client/scripts/createVersionSource.js]: Fixed in 99e3c4d91fc6f7079a2509553b3e3977efd234d5

**Comment by michalrus** [comment]
Okay, pushed a fix, I’ll now do some manual testing :hourglass_flowing_sand:

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (mkazlauskas) [inline packages/cardano-services-client/scripts/createVersionSource.js:26]: Doesn't look like this would work when used like `previouslySupported.filter(sameMajorVersion(version))`
```suggestion
const sameMajorVersion = (a: string) => (b: string) => {
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/cardano-services-client/scripts/createVersionSource.js
Comment (michalrus) [reply packages/cardano-services-client/scripts/createVersionSource.js]: @mkazlauskas fixed in 29f24e50a834dfcfbbb8b632ee9957fcb3700833

**Comment by michalrus** [comment]
@mkazlauskas okay, it seems to work, but when running `yarn build:version`, I’m getting this diff (syntax only):

```diff
diff --git a/packages/cardano-services-client/src/version.ts b/packages/cardano-services-client/src/version.ts
index 7ca012effa..b829d9768a 100644
--- a/packages/cardano-services-client/src/version.ts
+++ b/packages/cardano-services-client/src/version.ts
@@ -1,12 +1,12 @@
 // auto-generated using ../scripts/createVersionSource.js
 export const apiVersion = {
-  assetInfo: '1.0.0',
-  chainHistory: '3.0.1',
-  handle: '1.0.0',
-  networkInfo: '1.0.0',
-  rewards: '1.0.0',
-  root: '1.0.0',
-  stakePool: '1.1.0',
-  txSubmit: '2.0.0',
-  utxo: '2.0.0'
+  "assetInfo": "1.0.0",
+  "chainHistory": "3.0.1",
+  "handle": "1.0.0",
+  "networkInfo": "1.0.0",
+  "rewards": "1.0.0",
+  "root": "1.0.0",
+  "stakePool": "1.1.0",
+  "txSubmit": "2.0.0",
+  "utxo": "2.0.0"
 };
```

Should the code be adjusted to output an object in line with our linter config?

**Review by mkazlauskas — APPROVED** [review]
Looks good! :rocket: 
Approval assumes that some manual testing when bumping version in openApi.json was done

File: Dockerfile
Comment (michalrus) [reply Dockerfile:60]: @Daniele Ricci I [don’t know](https://input-output-rnd.slack.com/archives/C032DECBR6G/p1710338455751909?thread_ts=1710223845.458309&cid=C032DECBR6G) the answer, I only [tracked where the error occurred](https://github.com/input-output-hk/cardano-js-sdk/actions/runs/8264590398/job/22608506205?pr=1146#step:4:2222), and found a minimal solution.

I would like to understand this as well.

Comment (mkazlauskas) [reply Dockerfile:60]: That's odd. To start with, `cardano-services-client/test` should already exist :thinking:

Comment (iccicci) [reply Dockerfile:60]: All tests files are excluded through `.dockerignore` as they are (should be) not required in the docker image...

What I can't understand is: what else is changed in this PR to make this file required? Before this change it wasn't...

More curiosity than a reason to block this PR.

Comment (mkazlauskas) [reply Dockerfile:60]: Ah I see, .dockerignore :) We don't understand what changed to cause this, I think we should put some effort to figure it out

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
Nice work! Please squash all but 4faee1e26a8bad33a47aebfaf0a2a0cdcad4a1ca before merging

Also, `fix` commit type should only be used for code (application/library) fixes, because it will show up in the CHANGELOG. Please use `chore/build/ci`

**Review by iccicci — APPROVED** [review]
Good job! 🚀

**Comment by michalrus** [comment]
@mkazlauskas [done](https://github.com/input-output-hk/cardano-js-sdk/compare/78fed648ee69a045888a4d197a1e2d5725b5ba7d..c086a00544138593a0580225684eb0ea1e677e4d) :)

**Review by mkazlauskas — APPROVED** [review]
Nice work! :rocket:

**Pull request merged**
